feat: auto-surrend phase based

Dampen59 opened the pull request:
Added a auto-surrend feature after a specified amount of time (configurable in config)
Recommended absolute minimum is 600 secs (not code-locked for dev purposes)
Recommended running value : at least 700 secs (default is 720)
Feature is disabled by default
Bumped YAML version to 3
Added a send_key method to ADB (more reliable to access settings, I have joined a link if later we do an enum)
Added config entries for surrend feature
Added some output when in DEBUG mode

This have been tested for 3 games with no issues.
The bot automaticalluy queues again after surrender.

Kyrluckechuck reviewed (commented):
Biggest concern I have with this approach is that a set amount of time is a finicky/unreliable way to ensure the game has reached the "able to surrender" threshold, which is typically a set round number (3-2 for example), and in the previous iteration of this bot that most of this came from https://github.com/Kyrluckechuck/TFT-Bot, we approached that by doing the round matching.

@akshualy thoughts?

akshualy commented:
Thanks for taking your time to review as I'm out of country @Kyrluckechuck , I agree with you. The ideal approach would be to expand the timer at the top and check which major round we're on. (it contracts again after a while so there needs to be a check for both states of the top round indicator)

Dampen59 commented:
Hello, alright I will do the necessary

Dampen59 commented:
rebased on main

Dampen59 commented:
Work is done, the surrend system is now Phase based.
Based upon reaching Phase 3-2

Removed old time-based surrend code
Added Expand top bar button (retract was removed because it isn't needed as it auto retracts when going to fight or return to board)
Added Image of the 3-2 phase
Renamed confiig entries
Kept the old "seconds" entry which now acts as a random delay if the user wants to

Locked by Pylint too-many-statements, maybe we should apply a disable flag on loop or take_game_decision nah ?

Kyrluckechuck commented on main.py:
```diff
@@ -135,6 +136,21 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
     await asyncio.sleep(1)
 
 
+async def surrend_game(adb_instance: ADB):
```
nit: Would you mind updating all references to `surrend` to `surrender`?

Kyrluckechuck commented on main.py:
```diff
@@ -280,6 +296,11 @@ async def loop(adb_instance: ADB, config: AluneConfig):
                 logger.info("Queue lock released, likely loading into game now.")
             case GameState.IN_GAME:
                 logger.info("App state is in game, looping decision making and waiting for the exit button.")
+                _random_delay = 0
+                # If the surrend option is enabled AND if a max random delay have been set, choose a random value..
+                if config.get_auto_surrend() and config.get_auto_surrend_random_delay() > 0:
+                    _random_delay = _random.randint(1, config.get_auto_surrend_random_delay())
+
```
This could likely be handled by the actual `get_auto_surrend_random_delay` function where it returns either 0 or the rand int between 1 and the delay, which would help make it a bit more readable

Kyrluckechuck commented on main.py:
```diff
@@ -288,6 +309,26 @@ async def loop(adb_instance: ADB, config: AluneConfig):
+                    # If the surrend feature is enabled, we expand the top bar to check if
+                    # we're in a surrend-possible phase
+                    if config.get_auto_surrend():
+                        await adb_instance.click_button(Button.expand_top_bar)
+                        await asyncio.sleep(1)
+                        is_phase_3_2 = screen.get_on_screen(screenshot, Image.PHASE_3_2_FULL)
+                        if is_phase_3_2:
+                            if _random_delay == 0:
+                                logger.info("We're in a phase we can surrend ! No delay have been set in config file.")
+                                logger.info("Surrendering now !")
+                                surrend_game(adb_instance)
+                            else:
+                                logger.info(
+                                    "We're in a phase we can surrend. A random delay have been set in config file."
+                                )
+                                logger.info("Surrendering in {_random_delay} second(s) !")
+                                surrend_timer = Timer(_random_delay, await surrend_game(adb_instance))
+                                surrend_timer.start()
```
This could be extracted into a function to increase readability.

It doesn't _need to_, but it could return true/false based on if it needs to break from the game loop (the break below this)

Dampen59 commented:
Applied requested changes, better code organization and refactor

Kyrluckechuck commented on alune/config.py:
```diff
@@ -122,20 +125,25 @@ def get_traits(self) -> list[images.Trait]:
+        """
+        return (
+            0
+            if self._config["auto_surrender_random_delay"] <= 0
```
nit: I'm a bit fuzzy on my syntax here, will this account for `null` / not present? 🤔

Dampen59 replied:
Hello you are right, I have made a small change in ``config.py`` to check if it's ``NoneType``
Also, I forgot to await the ``check_surrender_state()`` so I awaited it :)

Kyrluckechuck reviewed (approved):
Thanks for making those changes! Overall this looks great.

As long as @akshualy has no concerns when they are back, LGTM 🚀

akshualy commented on alune/config.py:
```diff
@@ -121,3 +124,26 @@ def get_traits(self) -> list[images.Trait]:
+            An random Integer between [1 and auto_surrender_random_delay]
+            Returns 0 if feature disabled or negative value.
+        """
+        return (
+            0
+            if self._config["auto_surrender_random_delay"] is None or self._config["auto_surrender_random_delay"] <= 0
+            else _random.randint(1, self._config["auto_surrender_random_delay"])
+        )
```
No point in an inline statement if it's longer and harder to read than just writing it out:

```py
delay_upper_bound = self._config["surrender_delay"] or 0
if delay_upper_bound <= 0:
    return 0
return _random.randint(1, delay_upper_bound)
```

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+    """
+    # If the surrender feature is enabled, we expand the top bar to check if
+    # we're in a surrend-possible phase
+    if config.get_auto_surrender():
```
Early return makes code more readable:

```py
if not config.should_surrender():
    return False
await ...
```

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+        await adb_instance.click_button(Button.expand_top_bar)
+        await asyncio.sleep(1)
+        is_phase_3_2 = screen.get_on_screen(screenshot, Image.PHASE_3_2_FULL)
+        if is_phase_3_2:
```
Same here about early return.

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
     await asyncio.sleep(1)
 
 
+async def check_surrender_state(adb_instance: ADB, screenshot: ndarray, config: AluneConfig):
```
This is missing the return type and documentation about what it returns.

akshualy replied:
Return type means `) -> bool:`

Dampen59 replied:
mb

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
 
+async def check_surrender_state(adb_instance: ADB, screenshot: ndarray, config: AluneConfig):
+    """
+    Check if we're able to surrend from the current game state
```
`surrender`, `.` missing

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+    Check if we're able to surrend from the current game state
+
+    Args:
+        adb_instance: The adb instance to process the surrend phase.
```
`surrender`

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+        screenshot: The current screen.
+        config: An instance of the alune config to use.
+    """
+    # If the surrender feature is enabled, we expand the top bar to check if
+    # we're in a surrend-possible phase
```
No need for useless comments.

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+        await asyncio.sleep(1)
+        is_phase_3_2 = screen.get_on_screen(screenshot, Image.PHASE_3_2_FULL)
+        if is_phase_3_2:
+            random_delay = config.get_auto_surrender_random_delay()
+            if random_delay == 0:
+                logger.info("We're in a phase we can surrend ! No delay have been set in config file.")
+                logger.info("Surrendering now !")
+                surrender_game(adb_instance)
+            else:
+                logger.info("We're in a phase we can surrend. A random delay have been set in config file.")
+                logger.info(f"Surrendering in {random_delay} second(s) !")
+                surrender_timer = Timer(random_delay, await surrender_game(adb_instance))
+                surrender_timer.start()
+            return True
```
Simpler:

```py
surrender_delay = config.get_surrender_delay()
logger.info(f"Surrendering the game in {surrender_delay} seconds.")
await asyncio.sleep(surrender_delay)
return True
```

This works fine because asyncio sleep is quite capable and skips 0 sleep lengths.
Also removes the need for threading, working in the async loop is better.

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+    Args:
+        adb_instance: The adb instance to process the surrend phase.
+    """
+    await adb_instance.send_key(111)  # Send escape key, this open the settings menu
+    await asyncio.sleep(2)
+    await adb_instance.click_button(Button.surrender)  # Click the surrend button, this open a choice window
+    await asyncio.sleep(2)
+    await adb_instance.click_button(Button.check_surrender)  # Confirm surrend
+    await asyncio.sleep(5)
```
Please remove the comments here. The code is self-documenting - "click button ( surrender )".

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+
+async def surrender_game(adb_instance: ADB):
+    """
+    Surrends the current game
```
`Surrenders`, `.` missing

akshualy commented on main.py:
```diff
@@ -135,6 +136,51 @@ async def handle_augments(screenshot: ndarray, adb_instance: ADB):
+    Surrends the current game
+
+    Args:
+        adb_instance: The adb instance to process the surrend phase.
```
`surrender`.

akshualy reviewed (changes requested):
A lot of nitpicking, some needed but easy fixes.
Feel free to decline any nitpick, I just have a very specific style of what I consider clean code.

akshualy commented on alune/config.py:
```diff
@@ -121,3 +124,26 @@ def get_traits(self) -> list[images.Trait]:
+        Returns:
+            A Boolean that determines whether or not we should surrend when possible.
+        """
+        return self._config["auto_surrender"]
```
`surrender_early` is even better, good suggestion :)

akshualy reviewed (approved):
lgtm
